hnn_core/network.py

```diff
@@ -1187,6 +1187,160 @@ def add_connection(self, src_gids, target_gids, loc, receptor,
+        for target_gid in target_gids:
+            _validate_type(target_gid, int, 'target_gid')
+            gid_type = self.gid_to_type(target_gid)
+            if gid_type is None:
+                raise AssertionError(
+                    f'target_gid {src_gid} not in net.gid_ranges')
+
+        # Convert loc to list
+        if loc is None:
+            loc = list()
+        elif isinstance(loc, str):
+            loc = [loc]
+        for loc_item in loc:
+            _check_option('loc', loc_item, valid_loc)
+
+        # Convert receptor to list
+        if receptor is None:
+            receptor = list()
+        elif isinstance(receptor, str):
+            receptor = [receptor]
+        for receptor_item in receptor:
+            _check_option('receptor', receptor_item, valid_receptor)
+
+        # Create lookup dictionaries
+        src_dict, target_dict = dict(), dict()
+        loc_dict, receptor_dict = dict(), dict()
+        for conn_idx, conn in enumerate(self.connectivity):
+            # Store connections matching each src_gid
+            for src_gid in conn['src_gids']:
+                if src_gid in src_dict:
+                    src_dict[src_gid].append(conn_idx)
+                else:
+                    src_dict[src_gid] = [conn_idx]
+            # Store connections matching each target_gid
+            for target_gid in conn['target_gids']:
+                if target_gid in target_dict:
+                    target_dict[target_gid].append(conn_idx)
+                else:
+                    target_dict[target_gid] = [conn_idx]
+            # Store connections matching each location
+            if conn['loc'] in loc_dict:
+                loc_dict[conn['loc']].append(conn_idx)
+            else:
+                loc_dict[conn['loc']] = [conn_idx]
+            # Store connections matching each receptor
+            if conn['receptor'] in receptor_dict:
+                receptor_dict[conn['receptor']].append(conn_idx)
+            else:
+                receptor_dict[conn['receptor']] = [conn_idx]
+
+        # Look up conn indeces that match search terms and add to set.
+        conn_set = set()
+        search_pairs = [(src_gids, src_dict), (target_gids, target_dict),
+                        (loc, loc_dict), (receptor, receptor_dict)]
+        for search_terms, search_dict in search_pairs:
+            inner_set = set()
+            # Union of indices which match inputs for single parameter
+            for term in search_terms:
+                inner_set = inner_set.union(search_dict[term])
+            # Intersection across parameters
+            if conn_set and inner_set:
+                conn_set = conn_set.intersection(inner_set)
+                if not conn_set:  # Exit if both are non empty disjoint sets
+                    raise Warning('No connections match search parameters.'
+                                  'Try changing or excluding parameters.')
```

Comment:
I would just return an empty list without complaining. Warnings can be very notorious in terms of polluting the output log, specially if the user is running something in a loop and this condition happens often